chain/chunks/src/lib.rs

```diff
@@ -1568,13 +1568,15 @@ impl ShardsManager {
     }
 
     /// Inserts the header if it is not already known, and process the forwarded chunk parts cached
-    /// for this chunk, if any.
+    /// for this chunk, if any. Returns true if the header was newly inserted or forwarded parts
+    /// were newly processed.
     pub fn insert_header_if_not_exists_and_process_cached_chunk_forwards(
         &mut self,
         header: &ShardChunkHeader,
-    ) {
+    ) -> bool {
+        let header_known_before = self.encoded_chunks.get(&header.chunk_hash()).is_some();
```

Comment:
Should you return here if header_known_before is true?

Reply:
No because this function is also called in the process_partial_encoded_chunk case in which case we have the header but we may still need to apply forwards.